<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>DNA evolutionary models · PhyloTraits.jl</title><meta name="title" content="DNA evolutionary models · PhyloTraits.jl"/><meta property="og:title" content="DNA evolutionary models · PhyloTraits.jl"/><meta property="twitter:title" content="DNA evolutionary models · PhyloTraits.jl"/><meta name="description" content="Documentation for PhyloTraits.jl."/><meta property="og:description" content="Documentation for PhyloTraits.jl."/><meta property="twitter:description" content="Documentation for PhyloTraits.jl."/><meta property="og:url" content="https://JuliaPhylo.github.io/PhyloTraits.jl/man/fitdiscreteDNA/"/><meta property="twitter:url" content="https://JuliaPhylo.github.io/PhyloTraits.jl/man/fitdiscreteDNA/"/><link rel="canonical" href="https://JuliaPhylo.github.io/PhyloTraits.jl/man/fitdiscreteDNA/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="PhyloTraits.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="PhyloTraits.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PhyloTraits.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../phyloregression/">Continuous trait analysis</a></li><li><a class="tocitem" href="../simulate_continuous/">Continuous trait simulation</a></li><li><a class="tocitem" href="../fitDiscrete/">Discrete trait analysis</a></li><li class="is-active"><a class="tocitem" href>DNA evolutionary models</a><ul class="internal"><li><a class="tocitem" href="#reading-in-an-alignment"><span>reading in an alignment</span></a></li><li><a class="tocitem" href="#substitution-models-for-DNA"><span>substitution models for DNA</span></a></li><li><a class="tocitem" href="#likelihood-of-a-fixed-network"><span>likelihood of a fixed network</span></a></li></ul></li><li><a class="tocitem" href="../simulate_discrete/">Discrete trait simulation</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">public</a></li><li><a class="tocitem" href="../../lib/internal/">internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>DNA evolutionary models</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>DNA evolutionary models</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloTraits.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/main/docs/src/man/fitdiscreteDNA.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="fitting-DNA-substitution-parameters-on-a-network"><a class="docs-heading-anchor" href="#fitting-DNA-substitution-parameters-on-a-network">fitting DNA substitution parameters on a network</a><a id="fitting-DNA-substitution-parameters-on-a-network-1"></a><a class="docs-heading-anchor-permalink" href="#fitting-DNA-substitution-parameters-on-a-network" title="Permalink"></a></h1><p>The methods below model each DNA site as a trait, assuming that sites are unlinked, that is, they evolve independently of each other. In other words, this is a &quot;concatenation&quot; approach where sites from the same locus do not share information about their evolutionary path. This is appropriate if recombination is assumed to have occurred within genes.</p><h2 id="reading-in-an-alignment"><a class="docs-heading-anchor" href="#reading-in-an-alignment">reading in an alignment</a><a id="reading-in-an-alignment-1"></a><a class="docs-heading-anchor-permalink" href="#reading-in-an-alignment" title="Permalink"></a></h2><p>Like for trait evolution, <a href="../../lib/public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> can be used. It can be given data in a variety of ways. For DNA, this is one way, using <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.readfastatodna"><code>PhyloNetworks.readfastatodna</code></a>:</p><pre><code class="language-julia hljs"># read in network
dna_net = readnewick(&quot;((((((((((((((Ae_caudata_Tr275:1.0,Ae_caudata_Tr276:1.0):1.0,Ae_caudata_Tr139:1.0):1.0)#H1:1.0::0.6,((((((Ae_longissima_Tr241:1.0,Ae_longissima_Tr242:1.0):1.0,Ae_longissima_Tr355:1.0):1.0,(Ae_sharonensis_Tr265:1.0,Ae_sharonensis_Tr264:1.0):1.0):1.0,((Ae_bicornis_Tr408:1.0,Ae_bicornis_Tr407:1.0):1.0,Ae_bicornis_Tr406:1.0):1.0):1.0,((Ae_searsii_Tr164:1.0,Ae_searsii_Tr165:1.0):1.0,Ae_searsii_Tr161:1.0):1.0):1.0)#H2:1.0::0.6):1.0,(((Ae_umbellulata_Tr266:1.0,Ae_umbellulata_Tr257:1.0):1.0,Ae_umbellulata_Tr268:1.0):1.0,#H1:1.0::0.4):1.0):1.0,((Ae_comosa_Tr271:1.0,Ae_comosa_Tr272:1.0):1.0,(((Ae_uniaristata_Tr403:1.0,Ae_uniaristata_Tr357:1.0):1.0,Ae_uniaristata_Tr402:1.0):1.0,Ae_uniaristata_Tr404:1.0):1.0):1.0):1.0,(((Ae_tauschii_Tr352:1.0,Ae_tauschii_Tr351:1.0):1.0,(Ae_tauschii_Tr180:1.0,Ae_tauschii_Tr125:1.0):1.0):1.0,(#H2:1.0::0.4,((((Ae_mutica_Tr237:1.0,Ae_mutica_Tr329:1.0):1.0,Ae_mutica_Tr244:1.0):1.0,Ae_mutica_Tr332:1.0):1.0)#H4:1.0::0.6):1.0):1.0):1.0,(((T_boeoticum_TS8:1.0,(T_boeoticum_TS10:1.0,T_boeoticum_TS3:1.0):1.0):1.0,T_boeoticum_TS4:1.0):1.0,((T_urartu_Tr315:1.0,T_urartu_Tr232:1.0):1.0,(T_urartu_Tr317:1.0,T_urartu_Tr309:1.0):1.0):1.0):1.0):1.0,(((((Ae_speltoides_Tr320:1.0,Ae_speltoides_Tr323:1.0):1.0,Ae_speltoides_Tr223:1.0):1.0,Ae_speltoides_Tr251:1.0):1.0):1.0,#H4:1.0::0.4):1.0):1.0):1.0,Ta_caputMedusae_TB2:1.0):1.0,S_vavilovii_Tr279:1.0):1.0,Er_bonaepartis_TB1:1.0):1.0,H_vulgare_HVens23:1.0);&quot;);
# read in alignment in FASTA format
fastafile = joinpath(dirname(pathof(PhyloTraits)), &quot;..&quot;,&quot;examples&quot;,&quot;Ae_bicornis_Tr406_Contig10132.aln&quot;);
dna_dat, dna_weights = readfastatodna(fastafile, true);</code></pre><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; dna_dat</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">22×210 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> taxon                 x1   x2   x3   x4   x5   x6   x7   x8   x9   x10 </span> ⋯
     │<span class="sgr90"> String                DNA  DNA  DNA  DNA  DNA  DNA  DNA  DNA  DNA  DNA </span> ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Ae_uniaristata_Tr357  -    -    -    -    -    -    -    -    -    -    ⋯
   2 │ Ae_mutica_Tr237       -    -    -    -    -    -    -    -    -    -
   3 │ Ae_tauschii_Tr352     -    -    -    -    -    -    -    -    -    -
   4 │ Ae_searsii_Tr165      T    C    A    G    G    A    A    C    G    C
   5 │ Ae_sharonensis_Tr265  -    -    -    -    -    -    -    -    -    -    ⋯
   6 │ Ae_bicornis_Tr407     T    C    A    G    G    A    A    C    G    C
   7 │ Ae_longissima_Tr241   T    C    A    G    G    A    A    C    G    C
   8 │ Ae_bicornis_Tr406     T    C    A    G    G    A    A    C    G    C
  ⋮  │          ⋮             ⋮    ⋮    ⋮    ⋮    ⋮    ⋮    ⋮    ⋮    ⋮    ⋮   ⋱
  16 │ Ae_comosa_Tr271       -    -    -    -    -    -    -    -    -    C    ⋯
  17 │ Ae_searsii_Tr164      T    C    A    G    G    A    A    C    G    C
  18 │ Ae_speltoides_Tr251   -    -    -    -    -    -    -    -    -    -
  19 │ Ae_bicornis_Tr408     -    -    -    -    -    -    -    -    -    -
  20 │ Ae_tauschii_Tr351     T    C    A    G    G    G    A    C    A    C    ⋯
  21 │ Ae_longissima_Tr355   T    C    A    G    A    A    A    C    G    C
  22 │ Ae_caudata_Tr139      -    -    -    -    -    -    -    -    -    -
<span class="sgr36">                                                  199 columns and 7 rows omitted</span></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; dna_weights</code><code class="nohighlight hljs ansi" style="display:block;">209-element Vector{Float64}:
 23.0
 18.0
 13.0
 16.0
  1.0
  1.0
  1.0
  1.0
  1.0
  9.0
  ⋮
  1.0
  1.0
  1.0
  1.0
  1.0
  1.0
  1.0
  2.0
  3.0</code></pre><p>Here, <code>dna_dat</code> is a single data frame containing both species names and trait data (site patterns). The alignment was summarized by listing each observed site pattern only once in <code>dna_dat</code>. <code>dna_weights</code> is a vector of weights, containing the number of times that each site pattern was observed.</p><h2 id="substitution-models-for-DNA"><a class="docs-heading-anchor" href="#substitution-models-for-DNA">substitution models for DNA</a><a id="substitution-models-for-DNA-1"></a><a class="docs-heading-anchor-permalink" href="#substitution-models-for-DNA" title="Permalink"></a></h2><p>DNA-specific substitution models have 4 states: the 4 nucleotides from <a href="https://github.com/BioJulia/BioSymbols.jl">BioSymbols</a> (listed <a href="https://biojulia.dev/BioSymbols.jl/stable/nucleicacids/">here</a>). Each model has a relative and an absolute version.</p><ul><li><code>:JC69</code> Jukes &amp; Cantor 1969 model: one single rate for all transitions. The relative version has values -1 along the diagonal of the rate matrix (1 expected transition / unit of time). The absolute version has an extra parameter to scale the rate matrix.</li><li><code>:HKY85</code> Hasegawa, Kishino &amp; Yano 1985: treats transitions differently from transversions. The relative is scaled to predict an average of 1 transition / unit of time.</li></ul><p>We may allow for rate variation across sites using the <code>:RV</code> option.</p><h2 id="likelihood-of-a-fixed-network"><a class="docs-heading-anchor" href="#likelihood-of-a-fixed-network">likelihood of a fixed network</a><a id="likelihood-of-a-fixed-network-1"></a><a class="docs-heading-anchor-permalink" href="#likelihood-of-a-fixed-network" title="Permalink"></a></h2><p>In the examples below, none of the rate parameters are optimized, so we get to see the default starting values.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; d1 = fitdiscrete(dna_net, :JC69, dna_dat, dna_weights, :Gamma; optimizeQ=false, optimizeRVAS=false)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr33"><span class="sgr1">┌ Warning: </span></span>the network contains taxa with no data: those will be pruned
<span class="sgr33"><span class="sgr1">└ </span></span><span class="sgr90">@ PhyloTraits ~/work/PhyloTraits.jl/PhyloTraits.jl/src/fit_discrete.jl:921</span>
PhyloTraits.StatisticalSubstitutionModel:
Jukes and Cantor 69 Substitution Model,
  relative rate version
  off-diagonal rates equal to 1/3
  rate matrix Q:
                 A       C       G       T
         A       *  0.3333  0.3333  0.3333
         C  0.3333       *  0.3333  0.3333
         G  0.3333  0.3333       *  0.3333
         T  0.3333  0.3333  0.3333       *
Rate variation across sites: discretized Gamma
  alpha: 1.0
  categories for Gamma discretization: 4
  rates: [0.146, 0.513, 1.071, 2.27]
on a network with 3 reticulations
data:
  22 species
  1362 sites
  209 distinct patterns
log-likelihood: -14442.27642</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; d2 = fitdiscrete(dna_net, :HKY85, dna_dat, dna_weights, :Gamma; optimizeQ=false, optimizeRVAS=false)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr33"><span class="sgr1">┌ Warning: </span></span>the network contains taxa with no data: those will be pruned
<span class="sgr33"><span class="sgr1">└ </span></span><span class="sgr90">@ PhyloTraits ~/work/PhyloTraits.jl/PhyloTraits.jl/src/fit_discrete.jl:921</span>
PhyloTraits.StatisticalSubstitutionModel:
HKY85 Substitution Model base frequencies: [0.29100263035741913, 0.2682964567538295, 0.23638403218319665, 0.2043168807055547]
  relative rate version with transition/tranversion ratio kappa = 1.0,
   scaled so that there is one substitution per unit time
  rate matrix Q:
                 A       C       G       T
         A       *  0.3598  0.3170  0.2740
         C  0.3902       *  0.3170  0.2740
         G  0.3902  0.3598       *  0.2740
         T  0.3902  0.3598  0.3170       *
Rate variation across sites: discretized Gamma
  alpha: 1.0
  categories for Gamma discretization: 4
  rates: [0.146, 0.513, 1.071, 2.27]
on a network with 3 reticulations
data:
  22 species
  1362 sites
  209 distinct patterns
log-likelihood: -14420.83584</code></pre><p>When allowing for rate variation across sites with gamma-distributed rates, the default shape parameter α is 1.</p><p>In the more interesting examples below, we optimize the evolutionary rates and the way rates vary across sites (which is the default).</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; d3 = fitdiscrete(dna_net, :JC69, dna_dat, dna_weights, :Gamma; ftolAbs=0.1, xtolAbs=0.01)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr33"><span class="sgr1">┌ Warning: </span></span>the network contains taxa with no data: those will be pruned
<span class="sgr33"><span class="sgr1">└ </span></span><span class="sgr90">@ PhyloTraits ~/work/PhyloTraits.jl/PhyloTraits.jl/src/fit_discrete.jl:921</span>
PhyloTraits.StatisticalSubstitutionModel:
Jukes and Cantor 69 Substitution Model,
  relative rate version
  off-diagonal rates equal to 1/3
  rate matrix Q:
                 A       C       G       T
         A       *  0.3333  0.3333  0.3333
         C  0.3333       *  0.3333  0.3333
         G  0.3333  0.3333       *  0.3333
         T  0.3333  0.3333  0.3333       *
Rate variation across sites: discretized Gamma
  alpha: 0.05828
  categories for Gamma discretization: 4
  rates: [0.0, 0.0, 0.012, 3.988]
on a network with 3 reticulations
data:
  22 species
  1362 sites
  209 distinct patterns
log-likelihood: -3717.89988</code></pre><p>Lenient tolerance parameters <code>ftolAbs</code> etc. have been chosen here to make this example faster. Note that the fitted object contains a separate version of the input network, where any taxon without data has been pruned, and where branch numbers may have been modified.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; d3.net</code><code class="nohighlight hljs ansi" style="display:block;">PhyloNetworks.HybridNetwork, Rooted Network
53 edges
51 nodes: 22 tips, 3 hybrid nodes, 26 internal tree nodes.
tip labels: Ae_caudata_Tr275, Ae_caudata_Tr139, Ae_longissima_Tr241, Ae_longissima_Tr242, ...
(H_vulgare_HVens23:1.0,((((Ae_speltoides_Tr251:2.0):1.0,#H4:1.0::0.4):1.0,((((((Ae_caudata_Tr139:1.0,Ae_caudata_Tr275:2.0):1.0)#H1:1.0::0.6,((((((Ae_longissima_Tr241:1.0,Ae_longissima_Tr242:1.0):1.0,Ae_longissima_Tr355:1.0):1.0,Ae_sharonensis_Tr265:2.0):1.0,((Ae_bicornis_Tr408:1.0,Ae_bicornis_Tr407:1.0):1.0,Ae_bicornis_Tr406:1.0):1.0):1.0,(Ae_searsii_Tr164:1.0,Ae_searsii_Tr165:1.0):2.0):1.0)#H2:1.0::0.6):1.0,#H1:2.0::0.4):1.0,((Ae_comosa_Tr271:1.0,Ae_comosa_Tr272:1.0):1.0,((Ae_uniaristata_Tr403:1.0,Ae_uniaristata_Tr357:1.0):1.0,Ae_uniaristata_Tr402:1.0):2.0):1.0):1.0,(((Ae_tauschii_Tr352:1.0,Ae_tauschii_Tr351:1.0):1.0,Ae_tauschii_Tr125:2.0):1.0,(#H2:1.0::0.4,(Ae_mutica_Tr237:4.0)#H4:1.0::0.6):1.0):1.0):2.0):1.0):4.0);</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../fitDiscrete/">« Discrete trait analysis</a><a class="docs-footer-nextpage" href="../simulate_discrete/">Discrete trait simulation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Monday 13 January 2025 00:44">Monday 13 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
