<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>internals · PhyloTraits.jl</title><meta name="title" content="internals · PhyloTraits.jl"/><meta property="og:title" content="internals · PhyloTraits.jl"/><meta property="twitter:title" content="internals · PhyloTraits.jl"/><meta name="description" content="Documentation for PhyloTraits.jl."/><meta property="og:description" content="Documentation for PhyloTraits.jl."/><meta property="twitter:description" content="Documentation for PhyloTraits.jl."/><meta property="og:url" content="https://JuliaPhylo.github.io/PhyloTraits.jl/lib/internal/"/><meta property="twitter:url" content="https://JuliaPhylo.github.io/PhyloTraits.jl/lib/internal/"/><link rel="canonical" href="https://JuliaPhylo.github.io/PhyloTraits.jl/lib/internal/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="PhyloTraits.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="PhyloTraits.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PhyloTraits.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/phyloregression/">Continuous trait analysis</a></li><li><a class="tocitem" href="../../man/simulate_continuous/">Continuous trait simulation</a></li><li><a class="tocitem" href="../../man/fitDiscrete/">Discrete trait analysis</a></li><li><a class="tocitem" href="../../man/fitdiscreteDNA/">DNA evolutionary models</a></li><li><a class="tocitem" href="../../man/simulate_discrete/">Discrete trait simulation</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../man/example_fishes/">Xiphophorus fishes</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../public/">public</a></li><li class="is-active"><a class="tocitem" href>internals</a><ul class="internal"><li><a class="tocitem" href="#functions-and-types"><span>functions &amp; types</span></a></li><li><a class="tocitem" href="#index"><span>index</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>internals</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>internals</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloTraits.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/main/docs/src/lib/internal.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="internal-documentation"><a class="docs-heading-anchor" href="#internal-documentation">internal documentation</a><a id="internal-documentation-1"></a><a class="docs-heading-anchor-permalink" href="#internal-documentation" title="Permalink"></a></h1><p>Documentation for <code>PhyloTraits</code>&#39;s internal functions. These functions are not exported and their access (API) should not be considered stable. But they can still be used, like this for example: <code>PhyloTraits.foo()</code> for a function named <code>foo()</code>.</p><h2 id="functions-and-types"><a class="docs-heading-anchor" href="#functions-and-types">functions &amp; types</a><a id="functions-and-types-1"></a><a class="docs-heading-anchor-permalink" href="#functions-and-types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.BM" href="#PhyloTraits.BM"><code>PhyloTraits.BM</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">BM(λ)</code></pre><p>Brownian Motion, subtype of <a href="#PhyloTraits.ContinuousTraitEM"><code>ContinuousTraitEM</code></a>, to model the population mean of a trait (or of the residuals from a linear model). Under the BM model, the population (or species) means have a multivariate normal distribution with covariance matrix = σ²λV, where σ² is the between-species variance-rate (to be estimated), and the matrix V is obtained from <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.sharedpathmatrix-Tuple%7BHybridNetwork%7D"><code>PhyloNetworks.sharedpathmatrix</code></a><code>(net)[:tips]</code>.</p><p>λ is set to 1 by default, and is immutable. In future versions, λ may be used to control the scale for σ².</p><p>On a tree, V is the length of shared ancestry. On a network, the BM model assumes that the trait at a hybrid node is the weighted average of its immediate parents (plus possibly a fixed shift). The weights are the proportion of genes inherited from each parent: the γ parameters of hybrid edges.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L434-L452">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.ContinuousTraitEM" href="#PhyloTraits.ContinuousTraitEM"><code>PhyloTraits.ContinuousTraitEM</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ContinuousTraitEM</code></pre><p>Abstract type for evolutionary models for continuous traits, using a continuous-time stochastic process on a phylogeny.</p><p>For subtypes, see <a href="#PhyloTraits.BM"><code>BM</code></a>, <a href="#PhyloTraits.PagelLambda"><code>PagelLambda</code></a>, <a href="#PhyloTraits.ScalingHybrid"><code>ScalingHybrid</code></a>.</p><p>Each of these subtypes/models has the field <code>lambda</code>, whose default value is 1.0. However, the interpretation of this field differs across models.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L419-L429">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.NucleicAcidSubstitutionModel" href="#PhyloTraits.NucleicAcidSubstitutionModel"><code>PhyloTraits.NucleicAcidSubstitutionModel</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">NucleicAcidSubstitutionModel</code></pre><p>Adapted from <a href="https://github.com/BioJulia/SubstitutionModels.jl/">SubstitutionModels.jl</a> in BioJulia. The same <a href="../public/#PhyloTraits.Q-Tuple{BinaryTraitSubstitutionModel}"><code>Q</code></a> and <a href="#PhyloTraits.P-Tuple{PhyloTraits.SubstitutionModel, Float64}"><code>P</code></a> function names are used for the transition rates and probabilities.</p><p>For subtypes, see <a href="../public/#PhyloTraits.JC69"><code>JC69</code></a>, <a href="../public/#PhyloTraits.HKY85"><code>HKY85</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L31-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.OptSummary" href="#PhyloTraits.OptSummary"><code>PhyloTraits.OptSummary</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">OptSummary{T&lt;:AbstractFloat}</code></pre><p>Summary of an <code>NLopt</code> optimization. Idea and code taken from <a href="https://github.com/JuliaStats/MixedModels.jl"><code>MixedModels</code></a>. <code>T</code> is the type of the function argument(s) and of the function value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/nloptsummary.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.PagelLambda" href="#PhyloTraits.PagelLambda"><code>PhyloTraits.PagelLambda</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">PagelLambda(λ)</code></pre><p>Pagel&#39;s λ model, subtype of <a href="#PhyloTraits.ContinuousTraitEM"><code>ContinuousTraitEM</code></a>, with covariance matrix σ²V(λ). σ² is the between-species variance-rate (to be estimated), and V(λ) = λV + (1-λ)T, where V is the covariance under a Brownian motion <a href="#PhyloTraits.BM"><code>BM</code></a> and T is a diagonal matrix containing the total branch length elapsed from the root to each leaf (if the phylogeny is a tree, or more generally if the network is time consistent: the time from the root to a given node does not depend on the path).</p><p>λ ∈ [0,1] is mutable and may be optimized. It is a measure of phylogenetic signal, that is, how important the given network is for explaining variation in the response. When λ=1, the <code>PagelLambda</code> model reduces to the <code>BM</code> model.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L459-L472">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.ScalingHybrid" href="#PhyloTraits.ScalingHybrid"><code>PhyloTraits.ScalingHybrid</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ScalingHybrid(λ)</code></pre><p>Scaling Hybrid model, subtype of <a href="#PhyloTraits.ContinuousTraitEM"><code>ContinuousTraitEM</code></a>, with covariance matrix σ²V(N(λ)). σ² is the between-species variance-rate (to be estimated), V(N) is the Brownian motion <a href="#PhyloTraits.BM"><code>BM</code></a> covariance obtained from network N, and N(λ) is a obtained from the input network by rescaling the inheritance parameter γ of all minor edges by the same λ: a minor edge has its original γ changed to λγ, using the same λ at all reticulations. Note that for a major edge with original inheritance γ, the partner minor edge has inheritance γ<em>minor</em> = 1-γ, so the major edge&#39;s inheritance is changed to 1-λγ<em>minor</em> = λγ+1-λ. <em>Warning</em>: this model assumes that all hybrid nodes are bicombining, that is, they have exactly 2 parents each.</p><p>For more information: see Bastide (2017) dissertation, section 4.3.2 p.175, available at <a href="https://tel.archives-ouvertes.fr/tel-01629648">here</a>.</p><p>λ ∈ [0,λmax] is mutable and may be optimized. It is a measure of how important the reticulations are for explaining variation in the response.</p><ul><li>When λ = 1, the <code>ScalingHybrid</code> model reduces to the <code>BM</code> model on the original network N.</li><li>λ = 0 corresponds to the BM model on N&#39;s major tree (after removing all minor hybrid edges). Please note that testing the null hypothesis &quot;λ = 0&quot; would require comparing the likelihood ratio statistic to a null distribution that is <em>not</em> a chi-square distribution with 1 degree freedom, because 0 is at the boundary of allowable λ values. A mixture of 2 chi-square distributions might be appropriate (see for example <a href="https://doi.org/10.1214/19-EJS1576">Mitchell, Allman &amp; Rhodes 2019</a> and references therein).</li><li>λmax = 1/maximum(γ<em>minor</em>) ≥ 1. It is the maximum λ value such that λγ<em>minor</em> ≤ 1 at all hybrid nodes.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L479-L512">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.StatisticalSubstitutionModel" href="#PhyloTraits.StatisticalSubstitutionModel"><code>PhyloTraits.StatisticalSubstitutionModel</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StatisticalSubstitutionModel(
    model::SubstitutionModel,
    ratemodel::RateVariationAcrossSites,
    net::HybridNetwork,
    trait::AbstractVector,
    siteweight::Union{Nothing, Vector{Float64}}=nothing,
    maxhybrid::Int=length(net.hybrid)
)</code></pre><p>Inner constructor. Makes a deep copy of the input model, rate model. Warning: does <em>not</em> make a deep copy of the network: modification of the <code>object.net</code> would modify the input <code>net</code>. Assumes that the network has valid gamma values (to extract displayed trees).</p><pre><code class="nohighlight hljs">StatisticalSubstitutionModel(
    net::HybridNetwork,
    fastafile::String,
    modsymbol::Symbol,
    rvsymbol::Symbol=:noRV,
    ratecategories::Int=4;
    maxhybrid::Int=length(net.hybrid)
)</code></pre><p>Constructor from a network and a fasta file. The model symbol should be one of <code>:JC69</code>, <code>:HKY85</code>, <code>:ERSM</code> or <code>:BTSM</code>. The <code>rvsymbol</code> should be as required by <a href="../public/#PhyloTraits.RateVariationAcrossSites"><code>RateVariationAcrossSites</code></a>.</p><p>The network&#39;s gamma values are modified if they are missing. After that, a deep copy of the network is passed to the inner constructor.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L115-L145">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.StatisticalSubstitutionModel" href="#PhyloTraits.StatisticalSubstitutionModel"><code>PhyloTraits.StatisticalSubstitutionModel</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StatisticalSubstitutionModel</code></pre><p>Subtype of <code>StatsBase.StatisticalModel</code>, to fit discrete data to a model of trait substitution along a network. See <a href="../public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> to fit a trait substitution model to discrete data. It returns an object of type <code>StatisticalSubstitutionModel</code>, to which standard functions can be applied, like <code>loglikelihood(object)</code>, <code>aic(object)</code> etc.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L1-L9">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.SubstitutionModel" href="#PhyloTraits.SubstitutionModel"><code>PhyloTraits.SubstitutionModel</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">SubstitutionModel</code></pre><p>Abstract type for substitution models, using a continous time Markov model on a phylogeny. Adapted from <a href="https://github.com/BioJulia/SubstitutionModels.jl/">SubstitutionModels.jl</a> in BioJulia.</p><p>For variable rates, see <a href="../public/#PhyloTraits.RateVariationAcrossSites"><code>RateVariationAcrossSites</code></a></p><p>For sub types, see <a href="#PhyloTraits.NucleicAcidSubstitutionModel"><code>NucleicAcidSubstitutionModel</code></a>, <a href="../public/#PhyloTraits.TraitSubstitutionModel"><code>TraitSubstitutionModel</code></a></p><p>All these models are supposed to have fields <code>rate</code> and <code>eigeninfo</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1-L14">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.WithinSpeciesCTM" href="#PhyloTraits.WithinSpeciesCTM"><code>PhyloTraits.WithinSpeciesCTM</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">WithinSpeciesCTM</code></pre><p>Type to fit models accounting for within-species variation, including measurement error, genetic variation between individuals, plasticity, environmental variation etc. CTM stands for &quot;continuous trait model&quot;. Contains the estimated variance components (between-species phylogenetic variance rate and within-species variance) and output from the <code>NLopt</code> optimization used in the estimation.</p><p><strong>fields</strong></p><ul><li><code>wsp_var</code>: intra/within-species variance.</li><li><code>bsp_var</code>: inter/between-species variance-rate.</li><li><code>wsp_ninv</code>: vector of the inverse sample-sizes (e.g. [1/n₁, ..., 1/nₖ], where data from k species was used to fit the model and nᵢ is the no. of observations for the ith species).</li><li><code>rss</code>: within-species sum of squares</li><li><code>optsum</code>: an <a href="#PhyloTraits.OptSummary"><code>OptSummary</code></a> object.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L386-L405">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.getindex" href="#Base.getindex"><code>Base.getindex</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getindex(obj::TraitSimulation, d::Symbol, w::Symbol)</code></pre><p>Getting submatrices of an object of type <a href="../public/#PhyloTraits.TraitSimulation"><code>TraitSimulation</code></a>.</p><p><strong>Arguments</strong></p><ul><li><code>obj::TraitSimulation</code>: the matrix from which to extract.</li><li><code>d</code>: symbol specifying which sub-matrix to extract. Can be:<ul><li><code>:tips</code> columns and/or rows corresponding to the tips</li><li><code>:internalnodes</code> columns and/or rows corresponding to the internal nodes</li></ul></li><li><code>w</code>: symbol specifying whether simulated (<code>:sim</code>) or mean expected (<code>:exp</code>) values are desired.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/simulate_continuous.jl#L365-L377">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.P!-Tuple{AbstractMatrix, PhyloTraits.SubstitutionModel, Float64}" href="#PhyloTraits.P!-Tuple{AbstractMatrix, PhyloTraits.SubstitutionModel, Float64}"><code>PhyloTraits.P!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">P!(Pmat::AbstractMatrix, obj::SM, t::Float64)</code></pre><p>Fill in the input matrix <code>Pmat</code> with the transition rates to go from each state to another in time <code>t</code>, according to rates in <code>Q</code>. see also: <a href="#PhyloTraits.P-Tuple{PhyloTraits.SubstitutionModel, Float64}"><code>P</code></a>.</p><pre><code class="language-julia-repl hljs">julia&gt; m1 = BinaryTraitSubstitutionModel([1.0,2.0], [&quot;low&quot;,&quot;high&quot;])
Binary Trait Substitution Model:
rate low→high α=1.0
rate high→low β=2.0

julia&gt; PhyloTraits.P!(Matrix{Float64}(undef,2,2), m1, 0.3) # fills an uninitialized 2x2 matrix of floats
2×2 Matrix{Float64}:
 0.80219  0.19781
 0.39562  0.60438

julia&gt; m2 = JC69([1.]);

julia&gt; PhyloTraits.P!(Matrix{Float64}(undef,4,4), m2, 0.2)
4×4 Matrix{Float64}:
 0.824446   0.0585179  0.0585179  0.0585179
 0.0585179  0.824446   0.0585179  0.0585179
 0.0585179  0.0585179  0.824446   0.0585179
 0.0585179  0.0585179  0.0585179  0.824446 </code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L229-L256">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.P-Tuple{PhyloTraits.SubstitutionModel, Float64}" href="#PhyloTraits.P-Tuple{PhyloTraits.SubstitutionModel, Float64}"><code>PhyloTraits.P</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">P(obj, t)</code></pre><p>Probability transition matrix for a <a href="#PhyloTraits.SubstitutionModel"><code>SubstitutionModel</code></a>, of the form</p><pre><code class="nohighlight hljs">P[1,1] ... P[1,k]
   .          .
   .          .
P[k,1] ... P[k,k]</code></pre><p>where P[i,j] is the probability of ending in state j after time t, given that the process started in state i. see also: <a href="#PhyloTraits.P!-Tuple{AbstractMatrix, PhyloTraits.SubstitutionModel, Float64}"><code>P!</code></a>.</p><p>HKY example:</p><pre><code class="language-julia-repl hljs">julia&gt; m1 = HKY85([0.5], [0.20, 0.30, 0.30, 0.20])
HKY85 Substitution Model base frequencies: [0.2, 0.3, 0.3, 0.2]
relative rate version with transition/tranversion ratio kappa = 0.5,
 scaled so that there is one substitution per unit time
rate matrix Q:
               A       C       G       T
       A       *  0.4839  0.2419  0.3226
       C  0.3226       *  0.4839  0.1613
       G  0.1613  0.4839       *  0.3226
       T  0.3226  0.2419  0.4839       *

julia&gt; PhyloTraits.P(m1, 0.2)
4×4 StaticArraysCore.MMatrix{4, 4, Float64, 16} with indices SOneTo(4)×SOneTo(4):
 0.81592    0.0827167  0.0462192  0.0551445
 0.0551445  0.831326   0.0827167  0.0308128
 0.0308128  0.0827167  0.831326   0.0551445
 0.0551445  0.0462192  0.0827167  0.81592  </code></pre><p>Juke-Cantor example:</p><pre><code class="language-julia-repl hljs">julia&gt; m1 = JC69([1.]);

julia&gt; PhyloTraits.P(m1, 0.2)
4×4 StaticArraysCore.MMatrix{4, 4, Float64, 16} with indices SOneTo(4)×SOneTo(4):
 0.824446   0.0585179  0.0585179  0.0585179
 0.0585179  0.824446   0.0585179  0.0585179
 0.0585179  0.0585179  0.824446   0.0585179
 0.0585179  0.0585179  0.0585179  0.824446 </code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L175-L221">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.anova-Tuple{Vararg{PhyloNetworkLinearModel}}" href="#PhyloTraits.anova-Tuple{Vararg{PhyloNetworkLinearModel}}"><code>PhyloTraits.anova</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">anova(objs::PhyloNetworkLinearModel...)</code></pre><p>Takes several nested fits of the same data, and computes the F statistic for each pair of models.</p><p>The fits must be results of function <a href="../public/#PhyloTraits.phylolm"><code>phylolm</code></a> called on the same data, for models that have more and more effects.</p><p>Returns a DataFrame object with the anova table.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_phylolm_continuous.jl#L1112-L1122">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.check_matchtaxonnames!-Tuple{AbstractVector, AbstractVector, PhyloNetworks.HybridNetwork}" href="#PhyloTraits.check_matchtaxonnames!-Tuple{AbstractVector, AbstractVector, PhyloNetworks.HybridNetwork}"><code>PhyloTraits.check_matchtaxonnames!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">check_matchtaxonnames!(species, data, net; kwargs...)</code></pre><p>Modify <code>species</code> and <code>dat</code> by removing the species (rows) absent from the network. Return a new network (<code>net</code> is <em>not</em> modified) with tips matching those in species: if some species in <code>net</code> have no data, these species are pruned from the network. The network also has its node names reset, such that leaves have nodes have consecutive numbers starting at 1, with leaves first.</p><p>The optional keyword argument <code>suppresswarnings</code>, which is <code>false</code> by default, can be set to <code>true</code> to suppress warnings about taxa in the network without data.</p><p>Used by <a href="../public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> to build a new <a href="#PhyloTraits.StatisticalSubstitutionModel"><code>StatisticalSubstitutionModel</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L942-L955">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.defaultsubstitutionmodel" href="#PhyloTraits.defaultsubstitutionmodel"><code>PhyloTraits.defaultsubstitutionmodel</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">defaultsubstitutionmodel(network, modsymbol::Symbol, data::DataFrame,
              siteweights::Vector)</code></pre><p>Return a statistical substitution model (SSM) with appropriate state labels and a rate appropriate for the branch lengths in <code>net</code> (see <a href="#PhyloTraits.startingrate-Tuple{PhyloNetworks.HybridNetwork}"><code>startingrate</code></a>). The <code>data</code> frame must have the actual trait/site data in columns 2 and up, as when the species names are in column 1. For DNA data, the relative rate model is returned, with a stationary distribution equal to the empirical frequencies.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L1065-L1076">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.discrete_backwardlikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer}" href="#PhyloTraits.discrete_backwardlikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer}"><code>PhyloTraits.discrete_backwardlikelihood_trait!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">discrete_backwardlikelihood_trait!(obj::SSM, tree::Integer, ri::Integer)</code></pre><p>Update and return the backward likelihood (last argument <code>backwardlik</code>) assuming rate category <code>ri</code> and tree index <code>tree</code>, using current forward and backwards likelihoods in <code>obj</code>: these depend on the trait (or site) given to the last call to <code>discrete_corelikelihood_trait!</code>. Used by <code>ancestralreconstruction</code>.</p><p><strong>warning</strong>: assume correct transition probabilities.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_posterior_discrete.jl#L208-L219">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.discrete_corelikelihood!-Tuple{PhyloTraits.StatisticalSubstitutionModel}" href="#PhyloTraits.discrete_corelikelihood!-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>PhyloTraits.discrete_corelikelihood!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">discrete_corelikelihood!(obj::StatisticalSubstitutionModel;
                         whichtrait::AbstractVector{Int} = 1:obj.nsites)</code></pre><p>Calculate the likelihood and update <code>obj.loglik</code> for discrete characters on a network, calling <a href="#PhyloTraits.discrete_corelikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer, Integer}"><code>discrete_corelikelihood_trait!</code></a>. The algorithm extracts all displayed trees and weighs the likelihood under all these trees. The object&#39;s partial likelihoods are updated:</p><ul><li>forward and direct partial likelihoods are re-used, one trait at a time,</li><li>overall likelihoods on each displayed tree, given each rate category and for each given site/trait: are cached in <code>_loglikcache</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L771-L782">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.discrete_corelikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer, Integer}" href="#PhyloTraits.discrete_corelikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer, Integer}"><code>PhyloTraits.discrete_corelikelihood_trait!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">discrete_corelikelihood_trait!(obj::SSM, t::Integer, ci::Integer, ri::Integer)</code></pre><p>Return the likelihood for tree <code>t</code>, trait (character/site) index <code>ci</code> and rate category <code>ri</code>. Update &amp; modify the forward &amp; directional log-likelihoods <code>obj.forwardlik</code> and <code>obj.directlik</code>, which are indexed by [state, node<em>number or edge</em>number]. Used by <a href="#PhyloTraits.discrete_corelikelihood!-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>discrete_corelikelihood!</code></a>.</p><p><strong>Preconditions</strong>: <code>obj.logtrans</code> updated, edges directed, nodes/edges preordered</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L820-L829">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.formatinterval" href="#PhyloTraits.formatinterval"><code>PhyloTraits.formatinterval</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">formatinterval(obj::ReconstructedStates, pred::DataFrame, withexpectation::Bool=false, digits::Int=2)</code></pre><p>Format the prediction intervals for the plotting function. If <code>withexpectation</code> is set to true, then the best predicted value is also shown along with the interval.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_continuous.jl#L134-L140">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.getGammas-Tuple{PhyloNetworks.HybridNetwork}" href="#PhyloTraits.getGammas-Tuple{PhyloNetworks.HybridNetwork}"><code>PhyloTraits.getGammas</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getGammas(net)</code></pre><p>Vector of inheritance γ&#39;s of all major edges (tree edges and major hybrid edges), ordered according to the pre-order index of their child node, assuming this pre-order is already calculated (with up-to-date field <code>vec_node</code>). Here, a &quot;major&quot; edge is an edge with field <code>ismajor</code> set to true, regardless of its actual γ (below, at or above 0.5).</p><p>See <a href="#PhyloTraits.setGammas!-Tuple{PhyloNetworks.HybridNetwork, Vector}"><code>setGammas!</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_phylolm_continuous.jl#L143-L154">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.getmissingtipmarks" href="#PhyloTraits.getmissingtipmarks"><code>PhyloTraits.getmissingtipmarks</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getmissingtipmarks(obj::ReconstructedStates, missingmark::AbstractString=&quot;*&quot;)</code></pre><p>Create a vector of string, with a <code>missingmark</code> for tips that are missing.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_continuous.jl#L89-L93">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.getparameters-Tuple{PhyloTraits.RVASInv}" href="#PhyloTraits.getparameters-Tuple{PhyloTraits.RVASInv}"><code>PhyloTraits.getparameters</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getparameters(obj::RateVariationAcrossSites)</code></pre><p>Return a copy of the alpha and/or pinv parameters of model <code>obj</code>, in a single vector.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1153-L1158">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.getparamindex-Tuple{PhyloTraits.RVASInv}" href="#PhyloTraits.getparamindex-Tuple{PhyloTraits.RVASInv}"><code>PhyloTraits.getparamindex</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getparamindex(obj::RateVariationAcrossSites)</code></pre><p>Indices of parameters in (p_invariable, alpha).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1175-L1179">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.getpredint-Tuple{ReconstructedStates}" href="#PhyloTraits.getpredint-Tuple{ReconstructedStates}"><code>PhyloTraits.getpredint</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getpredint(obj::ReconstructedStates; level::Real=0.95)</code></pre><p>Prediction intervals with level <code>level</code> for internal nodes and missing tips.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_continuous.jl#L111-L115">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.lambda!-Tuple{PhyloNetworkLinearModel, Any}" href="#PhyloTraits.lambda!-Tuple{PhyloNetworkLinearModel, Any}"><code>PhyloTraits.lambda!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">lambda!(m::PhyloNetworkLinearModel, newlambda)
lambda!(m::ContinuousTraitEM, newlambda)</code></pre><p>Assign a new value to the lambda parameter.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L927-L932">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.lambda-Tuple{PhyloNetworkLinearModel}" href="#PhyloTraits.lambda-Tuple{PhyloNetworkLinearModel}"><code>PhyloTraits.lambda</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">lambda(m::PhyloNetworkLinearModel)
lambda(m::ContinuousTraitEM)</code></pre><p>Value assigned to the lambda parameter, if appropriate.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L918-L923">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.learnlabels-Tuple{Symbol, DataFrames.AbstractDataFrame}" href="#PhyloTraits.learnlabels-Tuple{Symbol, DataFrames.AbstractDataFrame}"><code>PhyloTraits.learnlabels</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">learnlabels(model::Symbol, dat::DataFrame)</code></pre><p>Return unique non-missing values in <code>dat</code>, and check that these labels can be used to construct of substitution model of type <code>model</code>.</p><p><strong>examples:</strong></p><pre><code class="language-julia-repl hljs">julia&gt; using DataFrames

julia&gt; dat = DataFrame(trait1 = [&quot;A&quot;, &quot;C&quot;, &quot;A&quot;, missing]); # 4×1 DataFrame

julia&gt; PhyloTraits.learnlabels(:BTSM, dat)
2-element Vector{String}:
 &quot;A&quot;
 &quot;C&quot;

julia&gt; PhyloTraits.learnlabels(:JC69, dat)
2-element Vector{String}:
 &quot;A&quot;
 &quot;C&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L1003-L1026">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.posterior_loghybridweight" href="#PhyloTraits.posterior_loghybridweight"><code>PhyloTraits.posterior_loghybridweight</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">posterior_loghybridweight(obj::SSM, hybrid_name, trait = 1)
posterior_loghybridweight(obj::SSM, edge_number, trait = 1)</code></pre><p>Log-posterior probability for all trees displaying the minor parent edge of hybrid node named <code>hybrid_name</code>, or displaying the edge number <code>edge_number</code>. That is: log of P(hybrid minor parent | trait) if a single <code>trait</code> is requested, or A[i]= log of P(hybrid minor parent | trait <code>i</code>) if <code>trait</code> is a vector or range (e.g. <code>trait = 1:obj.nsites</code>). These probabilities are conditional on the model parameters in <code>obj</code>.</p><p><strong>Precondition</strong>: <code>_loglikcache</code> updated by <a href="#PhyloTraits.discrete_corelikelihood!-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>discrete_corelikelihood!</code></a></p><p><strong>examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; net = readnewick(&quot;(((A:2.0,(B:1.0)#H1:0.1::0.9):1.5,(C:0.6,#H1:1.0::0.1):1.0):0.5,D:2.0);&quot;);

julia&gt; m1 = BinaryTraitSubstitutionModel([0.1, 0.1], [&quot;lo&quot;, &quot;hi&quot;]); # arbitrary rates

julia&gt; using DataFrames

julia&gt; dat = DataFrame(species=[&quot;C&quot;,&quot;A&quot;,&quot;B&quot;,&quot;D&quot;], trait=[&quot;hi&quot;,&quot;lo&quot;,&quot;lo&quot;,&quot;hi&quot;]);

julia&gt; fit = fitdiscrete(net, m1, dat); # optimized rates: α=0.27 and β=0.35

julia&gt; plhw = PhyloTraits.posterior_loghybridweight(fit, &quot;H1&quot;);

julia&gt; round(exp(plhw), digits=5) # posterior probability of going through minor hybrid edge
0.08017

julia&gt; hn = net.node[3]; getparentedgeminor(hn).gamma # prior probability
0.1</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_posterior_discrete.jl#L52-L86">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.posterior_logtreeweight" href="#PhyloTraits.posterior_logtreeweight"><code>PhyloTraits.posterior_logtreeweight</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">posterior_logtreeweight(obj::SSM, trait = 1)</code></pre><p>Array A of log-posterior probabilities for each tree displayed in the network: such that A[t] = log of P(tree <code>t</code> | trait <code>trait</code>) if a single <code>trait</code> is requested, or A[t,i]= log of P(tree <code>t</code> | trait <code>i</code>) if <code>trait</code> is a vector or range (e.g. <code>trait = 1:obj.nsites</code>). These probabilities are conditional on the model parameters in <code>obj</code>.</p><p>Displayed trees are listed in the order in which they are stored in the fitted model object <code>obj</code>.</p><p><strong>Precondition</strong>: <code>_loglikcache</code> updated by <a href="#PhyloTraits.discrete_corelikelihood!-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>discrete_corelikelihood!</code></a></p><p><strong>examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; net = readnewick(&quot;(((A:2.0,(B:1.0)#H1:0.1::0.9):1.5,(C:0.6,#H1:1.0::0.1):1.0):0.5,D:2.0);&quot;);

julia&gt; m1 = BinaryTraitSubstitutionModel([0.1, 0.1], [&quot;lo&quot;, &quot;hi&quot;]); # arbitrary rates

julia&gt; using DataFrames

julia&gt; dat = DataFrame(species=[&quot;C&quot;,&quot;A&quot;,&quot;B&quot;,&quot;D&quot;], trait=[&quot;hi&quot;,&quot;lo&quot;,&quot;lo&quot;,&quot;hi&quot;]);

julia&gt; fit = fitdiscrete(net, m1, dat); # optimized rates: α=0.27 and β=0.35

julia&gt; pltw = PhyloTraits.posterior_logtreeweight(fit);

julia&gt; round.(exp.(pltw), digits=5) # posterior trees probabilities (sum up to 1)
2-element Vector{Float64}:
 0.91983
 0.08017

julia&gt; round.(exp.(fit.priorltw), digits=4) # the prior tree probabilities are similar here (tiny data set!)
2-element Vector{Float64}:
 0.9
 0.1</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_posterior_discrete.jl#L1-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.setGammas!-Tuple{PhyloNetworks.HybridNetwork, Vector}" href="#PhyloTraits.setGammas!-Tuple{PhyloNetworks.HybridNetwork, Vector}"><code>PhyloTraits.setGammas!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setGammas!(net, γ vector)</code></pre><p>Set inheritance γ&#39;s of hybrid edges, using input vector for <em>major</em> edges. Assume pre-order calculated already, with up-to-date field <code>vec_node</code>. See <a href="#PhyloTraits.getGammas-Tuple{PhyloNetworks.HybridNetwork}"><code>getGammas</code></a>.</p><p><strong>Warning</strong>: very different from <a href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/lib/public/#PhyloNetworks.setgamma%21"><code>PhyloNetworks.setgamma!</code></a>, which focuses on a single hybrid event, updates the field <code>ismajor</code> according to the new γ, and is not used here.</p><p><strong>Assumption</strong>: each hybrid node has only 2 parents, a major and a minor parent (according to the edges&#39; field <code>ismajor</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_phylolm_continuous.jl#L118-L131">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.setalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGamma{S}, Float64}} where S" href="#PhyloTraits.setalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGamma{S}, Float64}} where S"><code>PhyloTraits.setalpha!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setalpha!(obj, alpha)</code></pre><p>Set the shape parameter <code>alpha</code> in a RateVariationAcrossSites model <code>obj</code>, and update the rate multipliers accordingly. Return the modified object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1026-L1032">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.seteigeninfo!-Tuple{BinaryTraitSubstitutionModel}" href="#PhyloTraits.seteigeninfo!-Tuple{BinaryTraitSubstitutionModel}"><code>PhyloTraits.seteigeninfo!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>for a [<code>BinaryTraitSubstitutionModel</code>]: store eigenvalue (q<em>01+q</em>10) and stationary distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L290-L292">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.seteigeninfo!-Tuple{EqualRatesSubstitutionModel}" href="#PhyloTraits.seteigeninfo!-Tuple{EqualRatesSubstitutionModel}"><code>PhyloTraits.seteigeninfo!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>for a [<code>EqualRatesSubstitutionModel</code>]: store lambda = k/(k-1), where k is the number of states</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L458-L460">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.seteigeninfo!-Tuple{HKY85}" href="#PhyloTraits.seteigeninfo!-Tuple{HKY85}"><code>PhyloTraits.seteigeninfo!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>for <a href="../public/#PhyloTraits.HKY85"><code>HKY85</code></a>: store piR, piY, the 2 non-zero eigenvalues and a scaling factor</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L651-L653">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.seteigeninfo!-Tuple{JC69}" href="#PhyloTraits.seteigeninfo!-Tuple{JC69}"><code>PhyloTraits.seteigeninfo!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>for <a href="../public/#PhyloTraits.JC69"><code>JC69</code></a>: store lambda = 4/3 (if relative) or rate * 4/3 (absolute).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L553-L555">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.seteigeninfo!-Tuple{PhyloTraits.SubstitutionModel}" href="#PhyloTraits.seteigeninfo!-Tuple{PhyloTraits.SubstitutionModel}"><code>PhyloTraits.seteigeninfo!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">seteigeninfo!(obj)</code></pre><p>Calculate eigenvalue &amp; eigenfector information for a substitution model (SM) object (as needed to calculate transition rate matrices) and store this info within the object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L61-L66">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.setparameters!-Tuple{PhyloTraits.RVASInv, AbstractVector}" href="#PhyloTraits.setparameters!-Tuple{PhyloTraits.RVASInv, AbstractVector}"><code>PhyloTraits.setparameters!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setparameters!(obj::RateVariationAcrossSites, par::AbstractVector)</code></pre><p>Set the values of the alpha and/or pinv parameters of model <code>obj</code>. See also <a href="#PhyloTraits.setalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGamma{S}, Float64}} where S"><code>setalpha!</code></a>, <a href="#PhyloTraits.setpinv!-Tuple{PhyloTraits.RVASInv, Float64}"><code>setpinv!</code></a> and <a href="#PhyloTraits.setpinvalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGammaInv{S}, Float64, Float64}} where S"><code>setpinvalpha!</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1163-L1168">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.setpinv!-Tuple{PhyloTraits.RVASInv, Float64}" href="#PhyloTraits.setpinv!-Tuple{PhyloTraits.RVASInv, Float64}"><code>PhyloTraits.setpinv!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setpinv!(obj, pinv)</code></pre><p>Set the proportion of invariable sites <code>pinv</code> in a RateVariationAcrossSites model <code>obj</code>, and update the rate multipliers &amp; weights accordingly. For <code>RVASInvGamma</code> objects, the original rate multipliers are assumed correct, according to the original <code>pinv</code> value. Return the modified object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1060-L1068">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.setpinvalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGammaInv{S}, Float64, Float64}} where S" href="#PhyloTraits.setpinvalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGammaInv{S}, Float64, Float64}} where S"><code>PhyloTraits.setpinvalpha!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setpinvalpha!(obj, pinv, alpha)</code></pre><p>Set the proportion of invariable sites <code>pinv</code> and the <code>alpha</code> parameter for the discretized gamma distribution in a model <code>obj</code> of type <code>RVASGammaInv{S}</code>. Update the rate multipliers &amp; weights accordingly. The mean of the distribution is constrained to 1.</p><p>Return the modified object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L1093-L1102">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.setrates!-Tuple{PhyloTraits.SubstitutionModel, AbstractVector}" href="#PhyloTraits.setrates!-Tuple{PhyloTraits.SubstitutionModel, AbstractVector}"><code>PhyloTraits.setrates!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setrates!(model, rates)</code></pre><p>update rates then call <a href="#PhyloTraits.seteigeninfo!-Tuple{BinaryTraitSubstitutionModel}"><code>seteigeninfo!</code></a> to update a model&#39;s eigeninfo</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L51-L55">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.showQ-Tuple{IO, PhyloTraits.SubstitutionModel}" href="#PhyloTraits.showQ-Tuple{IO, PhyloTraits.SubstitutionModel}"><code>PhyloTraits.showQ</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">showQ(IO, model)</code></pre><p>Print the Q matrix to the screen, with trait states as labels on rows and columns. adapted from prettyprint function by mcreel, found 2017/10 at https://discourse.julialang.org/t/display-of-arrays-with-row-and-column-names/1961/6</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_substitutions.jl#L146-L152">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.showdata" href="#PhyloTraits.showdata"><code>PhyloTraits.showdata</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">showdata(io::IO, obj::SSM, fullsiteinfo::Bool=false)</code></pre><p>Return information about the data in an SSM object: number of species, number or traits or sites, number of distinct patterns, and more information if <code>fullsiteinfo</code> is true: number sites with missing data only, number of invariant sites, number of sites with 2 distinct states, number of parsimony-informative sites (with 2+ states being observed in 2+ tips), number of sites with some missing data, and overall proportion of entries with missing data.</p><p>Note: Missing is not considered an additional state. For example, if a site contains some missing data, but all non-missing values take the same state, then this site is counted in the category &quot;invariant&quot;.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L189-L204">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.startingrate-Tuple{PhyloNetworks.HybridNetwork}" href="#PhyloTraits.startingrate-Tuple{PhyloNetworks.HybridNetwork}"><code>PhyloTraits.startingrate</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">startingrate(net)</code></pre><p>Estimate an evolutionary rate appropriate for the branch lengths in the network, which should be a good starting value before optimization in <code>fitdiscrete</code>, assuming approximately 1 change across the entire tree. If all edge lengths are missing, set starting rate to 1/(number of taxa).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L1044-L1051">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.traitlabels2indices-Tuple{AbstractVector, PhyloTraits.SubstitutionModel}" href="#PhyloTraits.traitlabels2indices-Tuple{AbstractVector, PhyloTraits.SubstitutionModel}"><code>PhyloTraits.traitlabels2indices</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">traitlabels2indices(data, model::SubstitutionModel)</code></pre><p>Check that the character states in <code>data</code> are compatible with (i.e. subset of) the trait labels in <code>model</code>. All columns are used. <code>data</code> can be a DataFrame or a Matrix (multiple traits), or a Vector (one trait). Return a vector of vectors (one per species) with integer entries, where each state (label) is replaced by its index in <code>model</code>. For DNA data, any ambiguous site is treated as missing.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L879-L888">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.update_logtrans-Tuple{PhyloTraits.StatisticalSubstitutionModel, PhyloNetworks.EdgeT{PhyloNetworks.Node}}" href="#PhyloTraits.update_logtrans-Tuple{PhyloTraits.StatisticalSubstitutionModel, PhyloNetworks.EdgeT{PhyloNetworks.Node}}"><code>PhyloTraits.update_logtrans</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">update_logtrans(obj::SSM, edge::Edge)</code></pre><p>Update the log-transition probabilities associates to one particular <code>edge</code> in the network.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L755-L760">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="PhyloTraits.update_logtrans-Tuple{PhyloTraits.StatisticalSubstitutionModel}" href="#PhyloTraits.update_logtrans-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>PhyloTraits.update_logtrans</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">update_logtrans(obj::SSM)</code></pre><p>Initialize and update <code>obj.logtrans</code>, the log transition probabilities along each edge in the full network. They are re-used for each displayed tree, which is why edges are not fused around degree-2 nodes when extracting displayed trees.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L734-L741">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.coeftable-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.coeftable-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.coeftable</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">coeftable(m::PhyloNetworkLinearModel; level::Real=0.95)</code></pre><p>Return coefficient estimates, standard errors, t-values, p-values, and t-intervals as a <code>StatsBase.CoefTable</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L750-L755">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.confint-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.confint-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.confint</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">confint(m::PhyloNetworkLinearModel; level::Real=0.95)</code></pre><p>Return confidence intervals for coefficients, with confidence level <code>level</code>, based on the t-distribution whose degree of freedom is determined by the number of species (as returned by <code>dof_residual</code>)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L737-L743">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.deviance" href="#StatsAPI.deviance"><code>StatsAPI.deviance</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StatsBase.deviance(m::PhyloNetworkLinearModel)</code></pre><p>-2 loglikelihood of the fitted model. See also  <a href="#StatsAPI.loglikelihood-Tuple{PhyloNetworkLinearModel}"><code>loglikelihood</code></a>.</p><p>Note: this is not the residual-sum-of-squares deviance as output by GLM, such as one would get with <code>deviance(m.model)</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L640-L647">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.deviance-Tuple{PhyloNetworkLinearModel, Val{true}}" href="#StatsAPI.deviance-Tuple{PhyloNetworkLinearModel, Val{true}}"><code>StatsAPI.deviance</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StatsBase.deviance(m::PhyloNetworkLinearModel, Val(true))</code></pre><p>Residual sum of squares with metric V, the estimated phylogenetic covariance, if the model is appropriate.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L651-L656">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.fit-Tuple{Type{PhyloTraits.StatisticalSubstitutionModel}, PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, RateVariationAcrossSites, AbstractVector}" href="#StatsAPI.fit-Tuple{Type{PhyloTraits.StatisticalSubstitutionModel}, PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, RateVariationAcrossSites, AbstractVector}"><code>StatsAPI.fit</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">fit(StatisticalSubstitutionModel, net, model, traits; kwargs...)
fit!(StatisticalSubstitutionModel; kwargs...)</code></pre><p>Internal function called by <a href="../public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a>: with same key word arguments <code>kwargs</code>. But dangerous: <code>traits</code> should be a vector of vectors as for <a href="../public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> <strong>but</strong> here <code>traits</code> need to contain the <em>indices</em> of trait values corresponding to the indices in <code>getlabels(model)</code>, and species should appear in <code>traits</code> in the order corresponding to the node numbers in <code>net</code>. See <a href="#PhyloTraits.traitlabels2indices-Tuple{AbstractVector, PhyloTraits.SubstitutionModel}"><code>traitlabels2indices</code></a> to convert trait labels to trait indices.</p><p><strong>Warning</strong>: does <em>not</em> perform checks. <a href="../public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> calls this function after doing checks, preordering nodes in the network, making sure nodes have consecutive numbers, species are matched between data and network etc.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_discrete.jl#L599-L613">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.loglikelihood-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.loglikelihood-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.loglikelihood</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loglikelihood(m::PhyloNetworkLinearModel)</code></pre><p>Log likelihood or log restricted likelihood (REML) depending on <code>m.reml</code>, of the fitted model.</p><p>For models with no within-species variation, the likelihood (or REML) is calculated based on the joint density for species-level mean responses.</p><p>For within-species variation models, the likelihood is calculated based on the joint density for individual-level responses. This can be calculated from individual-level data, but also by providing species-level means and standard deviations which is accepted by <a href="../public/#PhyloTraits.phylolm"><code>phylolm</code></a>.</p><p><strong>Warning</strong>: many summaries are based on the species-level model, like &quot;dof_residual&quot;, &quot;residuals&quot;, &quot;predict&quot; or &quot;deviance&quot;. So <code>deviance</code> is innapropriate to compare models with within-species variation. Use <code>loglikelihood</code> to compare models based on data at the individual level.</p><p>Reminder: do not compare ML or REML values across models fit on different data. Do not compare REML values across models that do not have the same predictors (fixed effects): use ML instead, for that purpose.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L784-L806">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.nobs-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.nobs-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.nobs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StatsBase.nobs(m::PhyloNetworkLinearModel)</code></pre><p>Number of observations: number of species with data, if the model assumes known species means, and number of individuals with data, if the model accounts for within-species variation.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L662-L668">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.nulldeviance-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.nulldeviance-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.nulldeviance</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StatsBase.nulldeviance(m::PhyloNetworkLinearModel)</code></pre><p>For appropriate phylogenetic linear models, the deviance of the null model  is the total sum of square with respect to the metric V, the estimated phylogenetic covariance matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L833-L839">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.predict-Tuple{ReconstructedStates}" href="#StatsAPI.predict-Tuple{ReconstructedStates}"><code>StatsAPI.predict</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">predict(
    obj::ReconstructedStates;
    interval::Union{Symbol,Nothing}=nothing,
    level::Real=0.95,
    text::Bool=false,
    digits::Int=2,
    missingmark::AbstractString=&quot;*&quot;,
    combine::Bool=false
)</code></pre><p>Estimated reconstructed states and, if <code>interval=:prediction</code>, prediction intervals with level <code>level</code> at all internal nodes and tips with missing traits.</p><p>If <code>text=true</code>, the prediction and intervals are formated as string for easy plotting with the <code>nodelabel</code> argument to <code>plot</code> from package <a href="https://github.com/juliaphylo/PhyloPlots.jl"><code>PhyloPlots</code></a>. In that case, <code>digits</code> controls the number of digits shown, <code>missingmark</code> adds a distinctive mark to the prediction of tips with missing data, (set to <code>missingmark=&quot;&quot;</code> for no mark), and if <code>combine=true</code>, the prediction and bound of the intervals are combined into a single text string.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/ancestral_continuous.jl#L31-L54">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.stderror-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.stderror-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.stderror</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">stderror(m::PhyloNetworkLinearModel)</code></pre><p>Return the standard errors of the coefficient estimates. See <a href="#StatsAPI.vcov-Tuple{PhyloNetworkLinearModel}"><code>vcov</code></a> for related information on how these are computed.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L727-L732">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsAPI.vcov-Tuple{PhyloNetworkLinearModel}" href="#StatsAPI.vcov-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.vcov</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">vcov(m::PhyloNetworkLinearModel)</code></pre><p>Return the variance-covariance matrix of the coefficient estimates.</p><p>For the continuous trait evolutionary models currently implemented, species-level mean response (conditional on the predictors), Y|X is modeled as:</p><ol><li>Y|X ∼ 𝒩(Xβ, σ²ₛV) for models assuming known species mean values (no within-species variation)</li><li>Y|X ∼ 𝒩(Xβ, σ²ₛV + σ²ₑD⁻¹) for models with information from multiple individuals and assuming within-species variation</li></ol><p>The matrix V is inferred from the phylogeny, but may also depend on additional parameters to be estimated (e.g. <code>lambda</code> for Pagel&#39;s Lambda model). See <a href="#PhyloTraits.ContinuousTraitEM"><code>ContinuousTraitEM</code></a>, <a href="../public/#PhyloTraits.PhyloNetworkLinearModel"><code>PhyloNetworkLinearModel</code></a> for more details.</p><p>If (1), then return σ²ₛ(X&#39;V⁻¹X)⁻¹, where σ²ₛ is estimated with REML, even if the model was fitted with <code>reml=false</code>. This follows the conventions of <a href="https://www.rdocumentation.org/packages/nlme/versions/3.1-152"><code>nlme::gls</code></a> and <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2"><code>stats::glm</code></a> in R.</p><p>If (2), then return σ²ₛ(X&#39;W⁻¹X)⁻¹, where W = V+(σ²ₑ/σ²ₛ)D⁻¹ is estimated, and σ²ₛ &amp; σₑ are the estimates obtained with ML or REML, depending on the <code>reml</code> option used to fit the model <code>m</code>. This follows the convention of <a href="https://juliastats.org/MixedModels.jl/stable/"><code>MixedModels.fit</code></a> in Julia.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/models_continuous.jl#L694-L719">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="StatsModels.isnested-Tuple{PhyloNetworkLinearModel, PhyloNetworkLinearModel}" href="#StatsModels.isnested-Tuple{PhyloNetworkLinearModel, PhyloNetworkLinearModel}"><code>StatsModels.isnested</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">isnested(m1::PhyloNetworkLinearModel, m2::PhyloNetworkLinearModel)
isnested(m1::ContinuousTraitEM, m2::ContinuousTraitEM)</code></pre><p>True if <code>m1</code> is nested in <code>m2</code>, false otherwise. Models fitted with different criteria (ML and REML) are not nested. Models with different predictors (fixed effects) must be fitted with ML to be considered nested.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/33023c37d4d6cc22020cf6117078fa227b32a232/src/fit_phylolm_continuous.jl#L1033-L1041">source</a></section></article><h2 id="index"><a class="docs-heading-anchor" href="#index">index</a><a id="index-1"></a><a class="docs-heading-anchor-permalink" href="#index" title="Permalink"></a></h2><ul><li><a href="#PhyloTraits.BM"><code>PhyloTraits.BM</code></a></li><li><a href="#PhyloTraits.ContinuousTraitEM"><code>PhyloTraits.ContinuousTraitEM</code></a></li><li><a href="#PhyloTraits.NucleicAcidSubstitutionModel"><code>PhyloTraits.NucleicAcidSubstitutionModel</code></a></li><li><a href="#PhyloTraits.OptSummary"><code>PhyloTraits.OptSummary</code></a></li><li><a href="#PhyloTraits.PagelLambda"><code>PhyloTraits.PagelLambda</code></a></li><li><a href="#PhyloTraits.ScalingHybrid"><code>PhyloTraits.ScalingHybrid</code></a></li><li><a href="#PhyloTraits.StatisticalSubstitutionModel"><code>PhyloTraits.StatisticalSubstitutionModel</code></a></li><li><a href="#PhyloTraits.StatisticalSubstitutionModel"><code>PhyloTraits.StatisticalSubstitutionModel</code></a></li><li><a href="#PhyloTraits.SubstitutionModel"><code>PhyloTraits.SubstitutionModel</code></a></li><li><a href="#PhyloTraits.WithinSpeciesCTM"><code>PhyloTraits.WithinSpeciesCTM</code></a></li><li><a href="#Base.getindex"><code>Base.getindex</code></a></li><li><a href="#PhyloTraits.P-Tuple{PhyloTraits.SubstitutionModel, Float64}"><code>PhyloTraits.P</code></a></li><li><a href="#PhyloTraits.P!-Tuple{AbstractMatrix, PhyloTraits.SubstitutionModel, Float64}"><code>PhyloTraits.P!</code></a></li><li><a href="#PhyloTraits.anova-Tuple{Vararg{PhyloNetworkLinearModel}}"><code>PhyloTraits.anova</code></a></li><li><a href="#PhyloTraits.check_matchtaxonnames!-Tuple{AbstractVector, AbstractVector, PhyloNetworks.HybridNetwork}"><code>PhyloTraits.check_matchtaxonnames!</code></a></li><li><a href="#PhyloTraits.defaultsubstitutionmodel"><code>PhyloTraits.defaultsubstitutionmodel</code></a></li><li><a href="#PhyloTraits.discrete_backwardlikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer}"><code>PhyloTraits.discrete_backwardlikelihood_trait!</code></a></li><li><a href="#PhyloTraits.discrete_corelikelihood!-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>PhyloTraits.discrete_corelikelihood!</code></a></li><li><a href="#PhyloTraits.discrete_corelikelihood_trait!-Tuple{PhyloTraits.StatisticalSubstitutionModel, Integer, Integer, Integer}"><code>PhyloTraits.discrete_corelikelihood_trait!</code></a></li><li><a href="#PhyloTraits.formatinterval"><code>PhyloTraits.formatinterval</code></a></li><li><a href="#PhyloTraits.getGammas-Tuple{PhyloNetworks.HybridNetwork}"><code>PhyloTraits.getGammas</code></a></li><li><a href="#PhyloTraits.getmissingtipmarks"><code>PhyloTraits.getmissingtipmarks</code></a></li><li><a href="#PhyloTraits.getparameters-Tuple{PhyloTraits.RVASInv}"><code>PhyloTraits.getparameters</code></a></li><li><a href="#PhyloTraits.getparamindex-Tuple{PhyloTraits.RVASInv}"><code>PhyloTraits.getparamindex</code></a></li><li><a href="#PhyloTraits.getpredint-Tuple{ReconstructedStates}"><code>PhyloTraits.getpredint</code></a></li><li><a href="#PhyloTraits.lambda-Tuple{PhyloNetworkLinearModel}"><code>PhyloTraits.lambda</code></a></li><li><a href="#PhyloTraits.lambda!-Tuple{PhyloNetworkLinearModel, Any}"><code>PhyloTraits.lambda!</code></a></li><li><a href="#PhyloTraits.learnlabels-Tuple{Symbol, DataFrames.AbstractDataFrame}"><code>PhyloTraits.learnlabels</code></a></li><li><a href="#PhyloTraits.posterior_loghybridweight"><code>PhyloTraits.posterior_loghybridweight</code></a></li><li><a href="#PhyloTraits.posterior_logtreeweight"><code>PhyloTraits.posterior_logtreeweight</code></a></li><li><a href="#PhyloTraits.setGammas!-Tuple{PhyloNetworks.HybridNetwork, Vector}"><code>PhyloTraits.setGammas!</code></a></li><li><a href="#PhyloTraits.setalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGamma{S}, Float64}} where S"><code>PhyloTraits.setalpha!</code></a></li><li><a href="#PhyloTraits.seteigeninfo!-Tuple{JC69}"><code>PhyloTraits.seteigeninfo!</code></a></li><li><a href="#PhyloTraits.seteigeninfo!-Tuple{HKY85}"><code>PhyloTraits.seteigeninfo!</code></a></li><li><a href="#PhyloTraits.seteigeninfo!-Tuple{BinaryTraitSubstitutionModel}"><code>PhyloTraits.seteigeninfo!</code></a></li><li><a href="#PhyloTraits.seteigeninfo!-Tuple{PhyloTraits.SubstitutionModel}"><code>PhyloTraits.seteigeninfo!</code></a></li><li><a href="#PhyloTraits.seteigeninfo!-Tuple{EqualRatesSubstitutionModel}"><code>PhyloTraits.seteigeninfo!</code></a></li><li><a href="#PhyloTraits.setparameters!-Tuple{PhyloTraits.RVASInv, AbstractVector}"><code>PhyloTraits.setparameters!</code></a></li><li><a href="#PhyloTraits.setpinv!-Tuple{PhyloTraits.RVASInv, Float64}"><code>PhyloTraits.setpinv!</code></a></li><li><a href="#PhyloTraits.setpinvalpha!-Union{Tuple{S}, Tuple{PhyloTraits.RVASGammaInv{S}, Float64, Float64}} where S"><code>PhyloTraits.setpinvalpha!</code></a></li><li><a href="#PhyloTraits.setrates!-Tuple{PhyloTraits.SubstitutionModel, AbstractVector}"><code>PhyloTraits.setrates!</code></a></li><li><a href="#PhyloTraits.showQ-Tuple{IO, PhyloTraits.SubstitutionModel}"><code>PhyloTraits.showQ</code></a></li><li><a href="#PhyloTraits.showdata"><code>PhyloTraits.showdata</code></a></li><li><a href="#PhyloTraits.startingrate-Tuple{PhyloNetworks.HybridNetwork}"><code>PhyloTraits.startingrate</code></a></li><li><a href="#PhyloTraits.traitlabels2indices-Tuple{AbstractVector, PhyloTraits.SubstitutionModel}"><code>PhyloTraits.traitlabels2indices</code></a></li><li><a href="#PhyloTraits.update_logtrans-Tuple{PhyloTraits.StatisticalSubstitutionModel}"><code>PhyloTraits.update_logtrans</code></a></li><li><a href="#PhyloTraits.update_logtrans-Tuple{PhyloTraits.StatisticalSubstitutionModel, PhyloNetworks.EdgeT{PhyloNetworks.Node}}"><code>PhyloTraits.update_logtrans</code></a></li><li><a href="#StatsAPI.coeftable-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.coeftable</code></a></li><li><a href="#StatsAPI.confint-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.confint</code></a></li><li><a href="#StatsAPI.deviance-Tuple{PhyloNetworkLinearModel, Val{true}}"><code>StatsAPI.deviance</code></a></li><li><a href="#StatsAPI.deviance"><code>StatsAPI.deviance</code></a></li><li><a href="#StatsAPI.fit-Tuple{Type{PhyloTraits.StatisticalSubstitutionModel}, PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, RateVariationAcrossSites, AbstractVector}"><code>StatsAPI.fit</code></a></li><li><a href="#StatsAPI.loglikelihood-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.loglikelihood</code></a></li><li><a href="#StatsAPI.nobs-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.nobs</code></a></li><li><a href="#StatsAPI.nulldeviance-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.nulldeviance</code></a></li><li><a href="#StatsAPI.predict-Tuple{ReconstructedStates}"><code>StatsAPI.predict</code></a></li><li><a href="#StatsAPI.stderror-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.stderror</code></a></li><li><a href="#StatsAPI.vcov-Tuple{PhyloNetworkLinearModel}"><code>StatsAPI.vcov</code></a></li><li><a href="#StatsModels.isnested-Tuple{PhyloNetworkLinearModel, PhyloNetworkLinearModel}"><code>StatsModels.isnested</code></a></li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../public/">« public</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 19 August 2025 18:30">Tuesday 19 August 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
