<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Discrete trait analysis · PhyloTraits.jl</title><meta name="title" content="Discrete trait analysis · PhyloTraits.jl"/><meta property="og:title" content="Discrete trait analysis · PhyloTraits.jl"/><meta property="twitter:title" content="Discrete trait analysis · PhyloTraits.jl"/><meta name="description" content="Documentation for PhyloTraits.jl."/><meta property="og:description" content="Documentation for PhyloTraits.jl."/><meta property="twitter:description" content="Documentation for PhyloTraits.jl."/><meta property="og:url" content="https://JuliaPhylo.github.io/PhyloTraits.jl/man/fitDiscrete/"/><meta property="twitter:url" content="https://JuliaPhylo.github.io/PhyloTraits.jl/man/fitDiscrete/"/><link rel="canonical" href="https://JuliaPhylo.github.io/PhyloTraits.jl/man/fitDiscrete/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="PhyloTraits.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="PhyloTraits.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PhyloTraits.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../phyloregression/">Continuous trait analysis</a></li><li><a class="tocitem" href="../simulate_continuous/">Continuous trait simulation</a></li><li class="is-active"><a class="tocitem" href>Discrete trait analysis</a><ul class="internal"><li><a class="tocitem" href="#Reading-trait-data"><span>Reading trait data</span></a></li><li><a class="tocitem" href="#Substitution-models"><span>Substitution models</span></a></li><li><a class="tocitem" href="#parameter-estimation-and-model-fit"><span>parameter estimation &amp; model fit</span></a></li><li><a class="tocitem" href="#ancestral-state-prediction"><span>ancestral state prediction</span></a></li><li><a class="tocitem" href="#impact-of-gene-flow-on-the-trait"><span>impact of gene flow on the trait</span></a></li></ul></li><li><a class="tocitem" href="../fitdiscreteDNA/">DNA evolutionary models</a></li><li><a class="tocitem" href="../simulate_discrete/">Discrete trait simulation</a></li></ul></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../example_fishes/">Xiphophorus fishes</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">public</a></li><li><a class="tocitem" href="../../lib/internal/">internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Discrete trait analysis</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Discrete trait analysis</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloTraits.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloTraits.jl/blob/main/docs/src/man/fitDiscrete.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Discrete-trait-evolution"><a class="docs-heading-anchor" href="#Discrete-trait-evolution">Discrete trait evolution</a><a id="Discrete-trait-evolution-1"></a><a class="docs-heading-anchor-permalink" href="#Discrete-trait-evolution" title="Permalink"></a></h1><p>With a phylogenetic network structure inferred, we can now estimate how quickly traits have evolved over time using a likelihood model. These traits should be discrete characteristics of a species such as feather color, diet type, or DNA in aligned genetic sequences.</p><p>As with continuous trait evolution, we assume a fixed network, correctly rooted, with branch lengths proportional to calendar time. We start with a network, then add data about the tips of this network.</p><h2 id="Reading-trait-data"><a class="docs-heading-anchor" href="#Reading-trait-data">Reading trait data</a><a id="Reading-trait-data-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-trait-data" title="Permalink"></a></h2><p>The simplest way is to use a vector of species names with a data frame of traits:</p><pre><code class="language-julia hljs"># read in network
net = readnewick(&quot;(O:4,(A:3,((B:0.4)#H1:1.6::0.92,((#H1:0::0.08,C:0.4):0.6,(D:.2,E:.2):0.8):1):1):1);&quot;);
# read in trait data
species = [&quot;C&quot;,&quot;A&quot;,&quot;D&quot;,&quot;B&quot;,&quot;O&quot;,&quot;E&quot;];
dat = DataFrame(trait=[&quot;hi&quot;,&quot;lo&quot;,&quot;lo&quot;,&quot;hi&quot;,&quot;lo&quot;,&quot;lo&quot;])</code></pre><p>If the species names and trait data are in the same data frame, read in your data frame then subset the data like this:</p><pre><code class="language-julia hljs">dat = DataFrame(
  species=[&quot;C&quot;,&quot;A&quot;,&quot;D&quot;,&quot;B&quot;,&quot;O&quot;,&quot;E&quot;],
  trait=[&quot;hi&quot;,&quot;lo&quot;,&quot;lo&quot;,&quot;hi&quot;,&quot;lo&quot;,&quot;lo&quot;]
);
species = dat.species # or: dat[!, :species]
select!(dat, Not(:species)) # select all columns except for :species; modifies dat in place</code></pre><p>Let&#39;s plot the network and map the data onto it:</p><pre><code class="language-julia hljs">using RCall, PhyloPlots
R&quot;par&quot;(mar=[0,0,0,0]); # to reduce margins
res = plot(net, tipoffset=0.3); # has point coordinates, for data annotation
o = indexin(tiplabels(net), species)
# o = 5,2,4,1,3,6 here: order to match taxa from &quot;species&quot; to tip labels
species[o] == tiplabels(net) # true :) sanity check
traitcolor = map(x -&gt; (x==&quot;lo&quot; ? &quot;grey&quot; : &quot;red&quot;), dat.trait[o])
ndf = res[:node_data] # node data frame: info for plotting nodes
leaves = ndf[!,:lea]
R&quot;points&quot;(x=ndf[leaves,:x] .+0.1, y=ndf[leaves,:y],
          pch=16, col=traitcolor, cex=1.5); # adds grey &amp; red points at tips
R&quot;legend&quot;(x=1, y=7, legend=[&quot;hi&quot;,&quot;lo&quot;], pch=16, col=[&quot;red&quot;,&quot;grey&quot;],
          title=&quot;my trait&quot;, bty=&quot;n&quot;,var&quot;title.adj&quot;=0);
# next: add to gene flow edge the proportion γ of genes affected
hi = findfirst([!e.ismajor for e in net.edge]) # hi = 6 : &quot;h&quot;ybrid &quot;i&quot;ndex
# hi = index of gene flow edge (minor hybrid) in net: horizontal segment
edf = res[:edge_data]
R&quot;text&quot;(edf[hi,:x]-0.3, edf[hi,:y]-0.1, edf[hi,:gam],
        col=&quot;deepskyblue&quot;, cex=0.75); # add the γ value</code></pre><p><img src="../../assets/figures/fitdiscrete_trait_net_1.svg" alt="net_1"/></p><h2 id="Substitution-models"><a class="docs-heading-anchor" href="#Substitution-models">Substitution models</a><a id="Substitution-models-1"></a><a class="docs-heading-anchor-permalink" href="#Substitution-models" title="Permalink"></a></h2><p>After reading in your data, we need to choose a model to describe how evolutionary changes (or substitutions, in the case of DNA) happened over time. Available Markov substitution models are described below.</p><p>For general trait types, use one of these three models:</p><ul><li><code>:BTSM</code> Binary Trait Substitution Model (2 states, rates unconstrained)</li><li><code>:ERSM</code> Equal Rates Substitution Model (<code>k</code> states, all transitions possible with equal rates)</li><li><code>:TBTSM</code> Two Binary Trait Substitution Model (though not fully implemented yet)</li></ul><p>To infer evolutionary rates, run <a href="../../lib/public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> on the network and data. It will calculate the maximum likelihood score of one or more discrete trait characters at the tips on a fixed network.</p><ul><li>Along each edge, evolutionary changes are modeled with a continous time Markov model.</li><li>At a hybrid node, the trait is assumed to be inherited from one or the other of its parents (immediately before the reticulation event), with probabilities equal to the inheritance γ of each parent edge, which is given by the network.</li><li>At the root of the network, a uniform distribution among the possible states is assumed a priori.</li><li>The model ignores incomplete lineage sorting (e.g. hemiplasy).</li></ul><h2 id="parameter-estimation-and-model-fit"><a class="docs-heading-anchor" href="#parameter-estimation-and-model-fit">parameter estimation &amp; model fit</a><a id="parameter-estimation-and-model-fit-1"></a><a class="docs-heading-anchor-permalink" href="#parameter-estimation-and-model-fit" title="Permalink"></a></h2><p>The example below if for a binary trait, first using a model assuming equal rates (from <code>lo</code> to <code>hi</code> and from <code>hi</code> to <code>lo</code>); then using a model allowing for distinct rates. The option <code>optimizeQ=false</code> causes transition rates to stay at their starting values, without being optimized.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; s1 = fitdiscrete(net, :ERSM, species, dat; optimizeQ=false)</code><code class="nohighlight hljs ansi" style="display:block;">PhyloTraits.StatisticalSubstitutionModel:
Equal Rates Substitution Model with k=2,
  all rates equal to α=0.07042.
  rate matrix Q:
                hi      lo
        hi       *  0.0704
        lo  0.0704       *
on a network with 1 reticulations
data:
  6 species
  1 trait
log-likelihood: -5.38599</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; s2 = fitdiscrete(net, :BTSM, species, dat; optimizeQ=false)</code><code class="nohighlight hljs ansi" style="display:block;">PhyloTraits.StatisticalSubstitutionModel:
Binary Trait Substitution Model:
  rate hi→lo α=0.07042
  rate lo→hi β=0.07042
on a network with 1 reticulations
data:
  6 species
  1 trait
log-likelihood: -5.38599</code></pre><p>The default rates, which act as starting value if rates were to be optimized, are chosen equal to the inverse of the total edge lengths in the network (or 1/number<em>of</em>taxa if all branch lengths are missing).</p><p>By default <code>optimizeQ = true</code>, such that <a href="../../lib/public/#PhyloTraits.fitdiscrete-Tuple{PhyloNetworks.HybridNetwork, PhyloTraits.SubstitutionModel, Dict}"><code>fitdiscrete</code></a> estimates the parameters of the rate matrix Q.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; s3 = fitdiscrete(net, :ERSM, species, dat)</code><code class="nohighlight hljs ansi" style="display:block;">PhyloTraits.StatisticalSubstitutionModel:
Equal Rates Substitution Model with k=2,
  all rates equal to α=0.70216.
  rate matrix Q:
                hi      lo
        hi       *  0.7022
        lo  0.7022       *
on a network with 1 reticulations
data:
  6 species
  1 trait
log-likelihood: -3.76738</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; s4 = fitdiscrete(net, :BTSM, species, dat)</code><code class="nohighlight hljs ansi" style="display:block;">PhyloTraits.StatisticalSubstitutionModel:
Binary Trait Substitution Model:
  rate hi→lo α=0.96783
  rate lo→hi β=0.55896
on a network with 1 reticulations
data:
  6 species
  1 trait
log-likelihood: -3.58702</code></pre><p>To compare the two models, we can use the Akaike criterion.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using StatsBase</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; aic(s3)</code><code class="nohighlight hljs ansi" style="display:block;">9.534758444197102</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; aic(s4)</code><code class="nohighlight hljs ansi" style="display:block;">11.174046169879334</code></pre><p>Here, the equal-rate model is slightly favored (lower AIC), so we will use <code>s3</code> below.</p><h2 id="ancestral-state-prediction"><a class="docs-heading-anchor" href="#ancestral-state-prediction">ancestral state prediction</a><a id="ancestral-state-prediction-1"></a><a class="docs-heading-anchor-permalink" href="#ancestral-state-prediction" title="Permalink"></a></h2><p>This is traditionally called &quot;ancestral state reconstruction&quot;, but we do not actually reconstruct anything. We make predictions for (past of present-day) values, hopefully with some measure to quantify our uncertainty.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; # show(ancestralreconstruction(s3), allrows=true)
       ancestralreconstruction(s3)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">13×4 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> nodenumber  nodelabel  hi         lo       </span>
     │<span class="sgr90"> Int64       String     Float64    Float64  </span>
─────┼────────────────────────────────────────────
   1 │          1  O          0.0        1.0
   2 │          2  A          0.0        1.0
   3 │          3  B          1.0        0.0
   4 │          4  C          1.0        0.0
   5 │          5  D          0.0        1.0
   6 │          6  E          0.0        1.0
   7 │          7  7          0.497566   0.502434
   8 │          8  8          0.497042   0.502958
   9 │          9  9          0.517993   0.482007
  10 │         10  10         0.476645   0.523355
  11 │         11  11         0.0226845  0.977316
  12 │         12  12         0.755186   0.244814
  13 │         13  H1         0.796886   0.203114</code></pre><p>Rows 1-6 correspond to the tips, with known values. We see much prediction uncertainty at most of the internal nodes. To see where these internal nodes (7-13/H1) are, we need to look at the network stored within the fitted model. This network might differ somewhat from the input network in case taxa with missing data where pruned, and with edges possibly renumbered.</p><pre><code class="language-julia hljs">plot(s3.net, shownodenumber=true, shownodelabel=true, tipoffset=0.2);</code></pre><p><img src="../../assets/figures/fitdiscrete_trait_net_2.svg" alt="net_2"/></p><p>Looking back at the posterior probabilities of states &quot;hi&quot; and &quot;lo&quot; at each node from the ancestral &#39;prediction&#39; table above, we see that there is more uncertainty near the root, and less uncertainty near the tips. The most recent common ancestor of D and E (node 11), in particular, is predicted to be &quot;lo&quot; with fairly high certainty.</p><h2 id="impact-of-gene-flow-on-the-trait"><a class="docs-heading-anchor" href="#impact-of-gene-flow-on-the-trait">impact of gene flow on the trait</a><a id="impact-of-gene-flow-on-the-trait-1"></a><a class="docs-heading-anchor-permalink" href="#impact-of-gene-flow-on-the-trait" title="Permalink"></a></h2><p>An interesting question is whether there is evidence that B obtained it&#39;s &quot;hi&quot; state via gene flow. The prior probability for this is γ: the supposedly known proportion of genes inherited via gene flow, which is part of the network (along with branch lengths). Here, this prior probability of trait inheritance via gene flow is:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; net.edge[6].gamma # the minor hybrid edge was edge 6, from above</code><code class="nohighlight hljs ansi" style="display:block;">1.0</code></pre><p>We can compare this to the posterior probability, and get a Bayes factor to compare the two hypotheses: gene flow vs. vertical inheritance.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; exp(s3.priorltw[1]) # prior: for vertical inheritance. &quot;ltw&quot; = log tree weight</code><code class="nohighlight hljs ansi" style="display:block;">0.92</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; exp(s3.priorltw[2]) # prior: for gene flow inheritance, same as γ above</code><code class="nohighlight hljs ansi" style="display:block;">0.07999999999999999</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; postltw = PhyloTraits.posterior_logtreeweight(s3)</code><code class="nohighlight hljs ansi" style="display:block;">2-element Vector{Float64}:
 -0.10479264829852397
 -2.3077104629296414</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; exp(postltw[2])     # posterior: for gene flow inheritance</code><code class="nohighlight hljs ansi" style="display:block;">0.09948877423615365</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; function geneflowBF(fit)
           postltw = PhyloTraits.posterior_logtreeweight(fit)
           exp(postltw[2] - postltw[1] + fit.priorltw[1] - fit.priorltw[2])
       end</code><code class="nohighlight hljs ansi" style="display:block;">geneflowBF (generic function with 1 method)</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; geneflowBF(s3)</code><code class="nohighlight hljs ansi" style="display:block;">1.2705237547097568</code></pre><p>We get a Bayes factor greater than 1, so there is more evidence that the &quot;hi&quot; value of B was inherited via gene flow, than via vertical inheritance. But the Bayes factor is just barely above 1, so the evidence is very equivocal. This may not be surprising given that gene flow occurred between fairly closely related species, and that the data set is very small.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../simulate_continuous/">« Continuous trait simulation</a><a class="docs-footer-nextpage" href="../fitdiscreteDNA/">DNA evolutionary models »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 19 August 2025 18:30">Tuesday 19 August 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
